internal/services/cragcast/cragcast.go

```diff
@@ -0,0 +1,227 @@
+	hours := []*HourlyForecast{}
+	// make a dictionary where times are the keys and values are hourly forecasts, loop through all sky cover, temp, precip etc...
+	// and add this info to each hour
+	// once that is done, loop through keys to make the hourly forecast and then sort by date
+
+	hoursMap := make(map[string]*HourlyForecast)
+
+	quantValuesToProcess := []QuantitativeValueType{TEMPERATURE, HUMIDITY, PRECIPITATION_PROBABILITY, QUANTITATIVE_PRECIPITATION, SKY_COVER, SNOWFALL_AMOUNT, WIND_DIRECTION, WIND_SPEED}
+
+	for _, quantValue := range quantValuesToProcess {
+		err := processQuantitativeValues(quantValue, weatherClientForecast.Properties.AdditionalProperties, hoursMap)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, hourlyForecast := range hoursMap {
+		hours = append(hours, hourlyForecast)
+	}
+
+	sort.Slice(hours, func(i, j int) bool {
+		return hours[i].StartOfHour.Before(hours[j].StartOfHour)
+	})
+
+	return hours, nil
+}
+
+type QuantitativeValueType string
+
+const (
+	TEMPERATURE                QuantitativeValueType = "temperature"
+	HUMIDITY                   QuantitativeValueType = "relativeHumidity"
+	QUANTITATIVE_PRECIPITATION QuantitativeValueType = "quantitativePrecipitation"
+	PRECIPITATION_PROBABILITY  QuantitativeValueType = "probabilityOfPrecipitation"
+	SKY_COVER                  QuantitativeValueType = "skyCover"
+	SNOWFALL_AMOUNT            QuantitativeValueType = "snowfallAmount"
+	WIND_DIRECTION             QuantitativeValueType = "windDirection"
+	WIND_SPEED                 QuantitativeValueType = "windSpeed"
+)
+
+func processQuantitativeValues(valueType QuantitativeValueType, additionalProperties map[string]noaaTypes.GridpointQuantitativeValueLayer, hoursMap map[string]*HourlyForecast) error {
+	var value noaaTypes.GridpointQuantitativeValueLayer
+
+	switch valueType {
+	case TEMPERATURE:
+		value = additionalProperties[string(TEMPERATURE)]
+	case HUMIDITY:
+		value = additionalProperties[string(HUMIDITY)]
+	case QUANTITATIVE_PRECIPITATION:
+		value = additionalProperties[string(QUANTITATIVE_PRECIPITATION)]
+	case PRECIPITATION_PROBABILITY:
+		value = additionalProperties[string(PRECIPITATION_PROBABILITY)]
+	case SKY_COVER:
+		value = additionalProperties[string(SKY_COVER)]
+	case SNOWFALL_AMOUNT:
+		value = additionalProperties[string(SNOWFALL_AMOUNT)]
+	case WIND_DIRECTION:
+		value = additionalProperties[string(WIND_DIRECTION)]
+	case WIND_SPEED:
+		value = additionalProperties[string(WIND_SPEED)]
+	default:
+		return errors.New("invalid value type")
+	}
+
+	for _, valueIter := range value.Values {
+		time, err := validTimeToTime(valueIter.ValidTime)
+		if err != nil {
+			return err
+		}
+		stringifiedTime := time.String()
+		hourObject := hoursMap[stringifiedTime]
+		if hourObject == nil {
+			hoursMap[stringifiedTime] = &HourlyForecast{
+				StartOfHour: *time,
+			}
+			hourObject = hoursMap[stringifiedTime]
+		}
+
+		switch valueType {
```

Comment:
duplicating these switch statements feels gross. Lemme know if you have a more elegant idea

Reply:
_still working on reasoning about the logic here, but:_

I recommended pulling this out to a single private function that can take whatever you want to mutate as an argument.